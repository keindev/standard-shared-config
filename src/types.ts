export enum FileType {
  Text = 'text',
  GLOB = 'glob',
  JSON = 'json',
  YAML = 'yaml',
}

export enum EntityName {
  Dependencies = 'dependencies',
  Scripts = 'scripts',
  Snapshots = 'snapshots',
}

export type IMergeRule = string[] | boolean;
export type IScript = [string, string];
export type IDependency = [string, string | undefined];

export interface ISnapshot {
  path: string;
  hash: string;
  merge: IMergeRule;
  type: FileType;
  content: string;
}
