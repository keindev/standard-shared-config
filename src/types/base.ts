export enum FileType {
  Text = 'text',
  GLOB = 'glob',
  JSON = 'json',
  YAML = 'yaml',
}

export enum EntityName {
  Scripts = 'scripts',
  Snapshots = 'snapshots',
  Package = 'package',
}

export type IMergeRule = string[] | null | boolean;
export type IScript = [string, string];
export type IMergeMap = { [key: string]: IMergeRule };

export interface ISnapshot {
  content: string;
  executable: boolean;
  hash: string;
  merge: IMergeRule;
  path: string;
  type: FileType;
}
