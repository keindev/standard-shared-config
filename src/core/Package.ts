import fs from 'fs';
import path from 'path';
import semver from 'semver';
import TaskTree from 'tasktree-cli';
import { PackageJson } from 'type-fest';
import writePkg from 'write-pkg';

import { IDependency, IScript } from '../types';

export default class Package {
  #data: PackageJson;

  constructor() {
    this.#data = JSON.parse(fs.readFileSync(path.resolve(process.cwd(), 'package.json'), 'utf8')) as PackageJson;
  }

  get name(): string {
    return path.basename(this.#data.name ?? '');
  }

  async update(): Promise<void> {
    this.#data.main = 'index.js';
    this.#data.bin = { [this.name]: `bin/${this.name}.js` };
    await this.write();
  }

  lint(dependencies: IDependency[]): void {
    if (dependencies.length) {
      const task = TaskTree.add('Lint package devDependencies:');

      dependencies.forEach(([name, version]) => {
        const currentVersion = semver.coerce((this.#data.devDependencies ?? {})[name]);
        const requiredVersion = semver.coerce(version);

        if (!currentVersion) {
          task.error(`{bold ${name}} is not presented!`, false);
        } else if (requiredVersion && semver.lt(currentVersion, requiredVersion)) {
          task.error(`${name} version is must be >=${version}!`, false);
        }
      });

      if (task.haveErrors) {
        task.fail('Some dependencies are missing or have wrong version:');
      } else {
        task.complete('All dependencies from shared config is presented in package.json');
      }
    }
  }

  async insert(scripts: IScript[]): Promise<void> {
    if (scripts.length) {
      const task = TaskTree.add('Inserting the required scripts:');
      const currentScripts = { ...(this.#data.scripts ?? {}) };

      scripts.forEach(([key, value]) => {
        if (!currentScripts[key]) task.log(`{bold ${key}} script added`);
        if (currentScripts[key] && currentScripts[key] !== value) task.log(`{bold ${key}} script replaced`);

        currentScripts[key] = value;
      });

      this.#data.scripts = Object.keys(currentScripts)
        .sort((a, b) => a.localeCompare(b))
        .reduce((acc, key) => ({ ...acc, [key]: currentScripts[key] }), {});
      await this.write();

      task.complete('Package scripts have been updated!');
    }
  }

  private async write(): Promise<void> {
    await writePkg({ ...(this.#data as { [key: string]: string }) });
  }
}
