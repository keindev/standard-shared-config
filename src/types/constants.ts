export const CONFIG_FILE = '.sharedconfig.yml';
export const SHARED_DIR = '.config';
export const OUTPUT_DIR = 'lib';
