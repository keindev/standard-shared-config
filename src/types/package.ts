import { ExportMap } from 'package-json-helper';
import { JSONValue } from 'package-json-helper/types/base';
import { ManagerType, PackageType } from 'package-json-helper/types/package';

export interface IPackageParams {
  exports?: JSONValue;
  manager?: ManagerType;
  peerDependencies?: Record<string, string>;
  type?: PackageType;
  types?: string;
}

export interface INormalizedPackageParams extends Pick<IPackageParams, 'manager' | 'type' | 'types'> {
  exports?: ExportMap;
  peerDependencies?: Record<string, string>;
}
