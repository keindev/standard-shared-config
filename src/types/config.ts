import { IMergeMap, IScript, ISnapshot } from './base.js';
import { INormalizedPackageParams, IPackageParams } from './package.js';

export interface ISharedConfig {
  executableFiles?: string[];
  ignorePatterns?: { [key: string]: string[] };
  include?: string[];
  mergeRules?: IMergeMap;
  outputDir?: string;
  package?: IPackageParams;
  scripts?: { [key: string]: string };
  sharedDir?: string;
}

export interface INormalizedSharedConfig {
  outputDir: string;
  package: IPackageParams;
  scripts: IScript[];
  sharedDir: string;
  snapshots: ISnapshot[];
}

export interface IExtractionConfig {
  overrideScripts?: { [key: string]: string };
}

export interface INormalizedExtractionConfig {
  package: INormalizedPackageParams;
  scripts: IScript[];
}
