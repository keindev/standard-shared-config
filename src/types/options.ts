import { EntityName, IScript, ISnapshot } from './base.js';
import { IPackageParams } from './package.js';

export interface IExtractionOptions {
  /** Package name */
  [EntityName.Name]?: string;
  /** Package parameters */
  [EntityName.Package]?: IPackageParams;
  /** List of package scripts */
  [EntityName.Scripts]?: IScript[];
  /** Configuration files snapshots */
  [EntityName.Snapshots]?: ISnapshot[];
}
