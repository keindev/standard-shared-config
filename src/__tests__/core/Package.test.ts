import fs from 'fs';
import stripAnsi from 'strip-ansi';
import { TaskTree } from 'tasktree-cli';
import { PackageJson } from 'type-fest';

import Package from '../../core/Package';
import { IDependency } from '../../types';

const PACKAGE_CONTENT = `
{
  "name": "@tagproject/standard-shared-config",
  "version": "0.1.0",
  "description": "One shared config to rule them all",
  "homepage": "https://github.com/keindev/standard-shared-config#readme",
  "license": "MIT",
  "devDependencies": {
    "typescript": "4.2.4"
  }
}
`;

let data: PackageJson = {};

jest.spyOn(fs, 'readFileSync').mockImplementation(() => PACKAGE_CONTENT);
jest.mock('write-pkg', (): ((value: PackageJson) => Promise<void>) => (value): Promise<void> => {
  data = value;

  return Promise.resolve();
});

describe('Package', () => {
  const pkg = new Package();

  it('initialization', () => {
    expect(pkg.name).toBe('standard-shared-config');
  });

  it('update', async () => {
    await pkg.update();

    expect(data?.main).toBe('index.js');
    expect(typeof data?.bin === 'object' ? data?.bin[pkg.name] : data?.bin).toBe(`bin/${pkg.name}`);
  });

  it('lint', () => {
    const task = TaskTree.tree();
    const errors: Error[] = [];
    const variations = [[['typescript', '4.x']], [['typescript', '5.x']]];

    variations.forEach(dependencies => {
      try {
        pkg.lint(dependencies as IDependency[]);
      } catch (error) {
        errors.push(error);
      }
    });

    expect(errors.length).toBe(1);
    expect(stripAnsi(task.render().join('\n'))).toMatchSnapshot();
  });
});
