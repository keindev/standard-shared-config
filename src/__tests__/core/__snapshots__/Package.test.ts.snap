// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Package lint 1`] = `
"  ✔ All dependencies from shared config is presented in package.json
  ✖ Some dependencies are missing or have wrong version: [fail]
  ─ ✖ typescript version is must be >=5.x!"
`;
